### PR TITLE
Add spiffetls.Listen* tests

### PR DESCRIPTION
**Pull request by amartinezfayo**
- Add spiffetls.Listen* tests
- Add tests for Close() (dial and listen)
- Add PeerID() testing

dial_test.go was renamed to spiffetls_test.go but git didn't capture that. If it's easier to review, I can open a PR with the changes applied to dial_test,go and then rename the file.

### CODE REVIEW

File: v2/spiffetls/spiffetls_test.go [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+			if test.listenLAddr == "" {
+				test.listenLAddr = defaultListenLAddr
+			}
+			if test.listenProtocol == "" {
+				test.listenProtocol = defaultListenProtocol
+			}
+			if test.listenMode != nil {
+				listener, err = spiffetls.ListenWithMode(listenCtx, test.listenProtocol, test.listenLAddr, test.listenMode, test.listenOption...)
+				if test.listenErr != "" {
+					require.Error(t, err)
```
Comment (azdagron) [inline v2/spiffetls/spiffetls_test.go:294]: nit: `EqualError` will cover the case where the err is nil so this isn't strictly necessary

File: v2/spiffetls/spiffetls_test.go [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+						require.Equal(t, clientID, spiffeID)
+					} else {
+						require.EqualError(t, err, test.serverConnPeerIDErr)
+					}
+				}()
+			} else {
+				// Test Listen function
+				listener, err = spiffetls.Listen(listenCtx, test.listenProtocol, test.listenLAddr, tlsconfig.AuthorizeID(clientID))
+				if test.listenErr != "" {
+					require.EqualError(t, err, test.listenErr)
```
Comment (marcosy) [inline v2/spiffetls/spiffetls_test.go:332]: We shouldn't call `require.*` or other testify packages that ends up calling to `t.FailNow()` inside go-routines.
```
FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test.
```
https://golang.org/pkg/testing/#T.FailNow

As an alternative we could exfiltrate the error through a channel and make the assertion in the `select` below.

Comment (azdagron) [reply v2/spiffetls/spiffetls_test.go:332]: Yes, good catch.

**Comment by azdagron** [comment]
I've been continuing to think about these tests since I reviewed. The body of the table driven test loop is quite large and somewhat hard to follow. Moving some of that body into helper functions might help. It may also make sense to have more than one table driven test function to reduce the optionality necessary to test everything in one. Removing the close testing into it's own table driven test for example would be one small simplification.

That being said, I'm ok taking them as-is (assuming marco's comment is addressed). I do think at some point we may want to refactor them a little.

File: v2/spiffetls/spiffetls_test.go (hunk 2 above)
Comment (amartinezfayo) [reply v2/spiffetls/spiffetls_test.go:332]: What if those `require.*` inside go routines are replaced with `assert.` calls? I went through them and continue the execution after the assertion failures shouldn't cause a panic. I'm just trying to think an alternative to not add extra complexity to the code.

Comment (azdagron) [reply v2/spiffetls/spiffetls_test.go:332]: Asserts from other goroutines are ok as long as they don't happen after the test method has finished (the testing package complains). So as long as the test waits for those goroutines to finish before returning, it should be alright.

**Comment by amartinezfayo** [comment]
@azdagron I've started writing the tests going down that path of trying to extract common code and split the testing in multiple table-driven tests and found that it was not that simple to extract the common pieces avoiding repetitive code to test dial and listen, so I gave it a try to see how a single table driven test would look like to test them. As you are pointing, it turned out that it ended up in a quite large loop and I'm concerned about the maintainability of this code.
If you agree, we can start with this implementation and I can work on improving it after this is merged in.

**Comment by azdagron** [comment]
Yep. Sounds good.
